File: task-manager/redis-tools/redis_group_query.py
#!/usr/bin/env python3
"""
Script to query all group information from Redis and display it.
"""

import redis
import json
import argparse
from typing import Dict, List, Any


def connect_to_redis(host='localhost', port=6379, db=0, password=None):
    """Connect to Redis instance."""
    try:
        r = redis.Redis(
            host=host,
            port=port,
            db=db,
            password=password,
            decode_responses=True
        )
        # Test connection
        r.ping()
        print(f"Connected to Redis at {host}:{port}")
        return r
    except Exception as e:
        print(f"Error connecting to Redis: {e}")
        return None


def get_all_keys(r, pattern="group:*"):
    """Get all keys matching the pattern."""
    try:
        keys = r.keys(pattern)
        return keys
    except Exception as e:
        print(f"Error getting keys: {e}")
        return []


def get_group_info(r, key):
    """Get information for a specific group key."""
    try:
        key_type = r.type(key)
        
        if key_type == 'string':
            value = r.get(key)
            return {'type': 'string', 'value': value}
        elif key_type == 'hash':
            value = r.hgetall(key)
            return {'type': 'hash', 'value': value}
        elif key_type == 'list':
            value = r.lrange(key, 0, -1)
            return {'type': 'list', 'value': value}
        elif key_type == 'set':
            value = r.smembers(key)
            return {'type': 'set', 'value': value}
        elif key_type == 'zset':
            value = r.zrange(key, 0, -1, withscores=True)
            return {'type': 'zset', 'value': value}
        else:
            return {'type': key_type, 'value': 'Unknown type'}
    except Exception as e:
        return {'error': str(e)}


def query_groups(r, patterns=["group:*", "groups:*", "*:group*", "*:groups*"]):
    """Query Redis for all group-related information."""
    all_groups = {}
    
    for pattern in patterns:
        print(f"\nSearching for keys matching pattern: {pattern}")
        keys = get_all_keys(r, pattern)
        
        if not keys:
            print(f"No keys found for pattern: {pattern}")
            continue
            
        print(f"Found {len(keys)} keys for pattern: {pattern}")
        
        for key in keys:
            group_info = get_group_info(r, key)
            all_groups[key] = group_info
    
    return all_groups


def display_group_info(groups):
    """Display group information in a formatted way."""
    if not groups:
        print("\nNo group information found in Redis.")
        return
    
    print(f"\n--- Found {len(groups)} group-related entries ---")
    
    for key, info in groups.items():
        print(f"\nKey: {key}")
        print(f"Type: {info['type']}")
        
        if 'error' in info:
            print(f"Error: {info['error']}")
        else:
            value = info['value']
            
            if isinstance(value, dict):
                print("Value:")
                for k, v in value.items():
                    print(f"  {k}: {v}")
            elif isinstance(value, list) or isinstance(value, set):
                print("Value:")
                for item in value:
                    print(f"  - {item}")
            elif isinstance(value, str):
                # Try to parse as JSON if it looks like it might be
                if value.startswith('{') or value.startswith('['):
                    try:
                        parsed_value = json.loads(value)
                        print("Value (JSON):")
                        print(json.dumps(parsed_value, indent=2))
                    except json.JSONDecodeError:
                        print(f"  Value: {value}")
                else:
                    print(f"  Value: {value}")
            else:
                print(f"  Value: {value}")
        
        print("-" * 40)


def main():
    parser = argparse.ArgumentParser(description='Query all group information from Redis')
    parser.add_argument('--host', default='localhost', help='Redis host (default: localhost)')
    parser.add_argument('--port', type=int, default=6379, help='Redis port (default: 6379)')
    parser.add_argument('--db', type=int, default=0, help='Redis database (default: 0)')
    parser.add_argument('--password', help='Redis password (if required)')
    parser.add_argument('--pattern', action='append', dest='patterns', 
                       help='Custom pattern to search for (can be used multiple times)')
    
    args = parser.parse_args()
    
    # Connect to Redis
    r = connect_to_redis(args.host, args.port, args.db, args.password)
    if not r:
        return
    
    # Set default patterns if none provided
    if args.patterns:
        patterns = args.patterns
    else:
        patterns = ["group:*", "groups:*", "*:group*", "*:groups*"]
    
    # Query for group information
    groups = query_groups(r, patterns)
    
    # Display the results
    display_group_info(groups)


if __name__ == "__main__":
    main()

File: task-manager/redis-tools/redis_group_query.ts
import { RedisClient } from './Task_Workflow\task-workflow-backend\src\core\repositories\RedisClient';

// Load environment variables
const env = {
  UPSTASH_REDIS_REST_URL: process.env.UPSTASH_REDIS_REST_URL!,
  UPSTASH_REDIS_REST_TOKEN: process.env.UPSTASH_REDIS_REST_TOKEN!,
};

async function queryAllGroups() {
  try {
    const redis = new RedisClient(env as any);
    
    // Get all keys that might contain group information
    // Look for patterns that might indicate group data
    const allKeys = await redis.list({ prefix: '*' });
    
    console.log('All Redis keys:');
    for (const keyObj of allKeys.keys) {
      console.log(`- ${keyObj.name}`);
    }
    
    // Look specifically for group-related keys
    const groupKeys = allKeys.keys.filter(key => 
      key.name.includes('group') || 
      key.name.includes('grp') ||
      key.name.startsWith('g_') ||
      key.name.startsWith('group_')
    );
    
    console.log('\nGroup-related keys:');
    for (const keyObj of groupKeys) {
      try {
        const value = await redis.get(keyObj.name);
        console.log(`\nKey: ${keyObj.name}`);
        console.log(`Value:`, value);
      } catch (err) {
        console.log(`Error reading key ${keyObj.name}:`, err.message);
      }
    }
    
    // If we don't find specific group patterns, try to get all values
    if (groupKeys.length === 0) {
      console.log('\nTrying to get common group storage patterns...');
      
      // Common patterns for group data
      const patterns = [
        'groups:*',
        'group:*',
        'groups_list',
        'all_groups',
        'team:*',
        'teams:*'
      ];
      
      for (const pattern of patterns) {
        try {
          const keys = await redis.list({ prefix: pattern });
          for (const keyObj of keys.keys) {
            const value = await redis.get(keyObj.name);
            console.log(`\nKey: ${keyObj.name}`);
            console.log(`Value:`, value);
          }
        } catch (err) {
          console.log(`Pattern ${pattern} not found or error:`, err.message);
        }
      }
    }
    
  } catch (error) {
    console.error('Error querying Redis:', error);
  }
}

// Run the query
queryAllGroups();